File: apps/ai-service/src/ranking/explain.py
from typing import Optional
from ..graph.base import GraphStore
from ..matching.coverage import RequirementMatch


def format_explanation(graph: GraphStore, match: RequirementMatch) -> str:
    """
    Generates a clear explanation for a single requirement match based on LCA.
    """
    r_node = graph.get_node(match.requirement_node_id)
    r_name = r_node.canonical_name if r_node else match.requirement_node_id

    if not match.best_user_skill_id:
        return f"Requirement '{r_name}': Unmet (No matching skills. Score: 0.0)"

    u_node = graph.get_node(match.best_user_skill_id)
    u_name = u_node.canonical_name if u_node else match.best_user_skill_id

    feats = match.structural_features
    if not feats:
        return f"Requirement '{r_name}': Matched by '{u_name}' (Score: {match.score})"

    if feats.exact_match:
        return f"Requirement '{r_name}': Direct exact match with '{u_name}' (Score: 1.0)"

    lca_node = graph.get_node(feats.lca) if feats.lca else None
    lca_name = lca_node.canonical_name if lca_node else (feats.lca or "General")

    if feats.user_is_descendant:
        return (
            f"Requirement '{r_name}': Specific satisfies broad — "
            f"Skill '{u_name}' is a specialized component of '{r_name}' "
            f"(Score: {match.score})"
        )

    if feats.user_is_ancestor:
        return (
            f"Requirement '{r_name}': Broad background — "
            f"Skill '{u_name}' is a broad parent category of specific requirement '{r_name}' "
            f"(Discounted Score: {match.score})"
        )

    if feats.same_parent:
        return (
            f"Requirement '{r_name}': Sibling technology — "
            f"Skill '{u_name}' shares the same parent '{lca_name}' as '{r_name}' "
            f"(Score: {match.score})"
        )

    if feats.lca_depth >= 2:
        return (
            f"Requirement '{r_name}': Related subdomain — "
            f"Skill '{u_name}' shares subdomain '{lca_name}' with '{r_name}' "
            f"(Score: {match.score})"
        )

    if feats.lca_depth == 1:
        return (
            f"Requirement '{r_name}': Broad domain overlap — "
            f"Skill '{u_name}' connected only at root domain level '{lca_name}' "
            f"(Score: {match.score})"
        )

    return f"Requirement '{r_name}': Weakly matched by '{u_name}' (Score: {match.score})"

File: apps/api/src/taxonomy/explain.ts
import { InMemoryTreeStore } from "./tree.store.js";
import { StructuralFeatures } from "./taxonomy.types.js";

export function formatExplanation(
  graph: InMemoryTreeStore,
  requirementNodeId: string,
  bestUserSkillId: string | null,
  score: number,
  features?: StructuralFeatures | null
): string {
  const rNode = graph.getNode(requirementNodeId);
  const rName = rNode?.canonical_name || requirementNodeId;
  const rDepth = rNode ? rNode.depth : (features?.requirement_depth ?? 0);

  if (!bestUserSkillId || score === 0 || !features || features.graph_distance >= 999) {
    return `Requirement '${rName}' (Depth: ${rDepth}): Unmet — No matching competencies found in candidate profile → Score: 0.00 (0%)`;
  }

  const uNode = graph.getNode(bestUserSkillId);
  const uName = uNode?.canonical_name || bestUserSkillId;
  const uDepth = uNode ? uNode.depth : (features.user_depth ?? 0);

  const lcaNode = features.lca ? graph.getNode(features.lca) : null;
  const lcaName = lcaNode?.canonical_name || features.lca || "Root";
  const lcaDepth = lcaNode ? lcaNode.depth : (features.lca_depth ?? 0);
  const distance = features.graph_distance;
  const scoreFormatted = `Score: ${score.toFixed(2)} (${Math.round(score * 100)}%)`;

  if (features.exact_match) {
    return `Requirement '${rName}' (Depth: ${rDepth}): Direct 1:1 exact match with verified skill '${uName}' (Depth: ${uDepth}) → Score: 1.00 (100%)`;
  }

  if (features.user_is_descendant) {
    return (
      `Requirement '${rName}' (Depth: ${rDepth}): Specific satisfies broad — ` +
      `Candidate skill '${uName}' (Depth: ${uDepth}) is a specialized component of '${rName}'. ` +
      `Intersection: '${lcaName}' (Depth: ${lcaDepth}, Graph Distance: ${distance}) → ${scoreFormatted}`
    );
  }

  if (features.user_is_ancestor) {
    return (
      `Requirement '${rName}' (Depth: ${rDepth}): Broad domain background — ` +
      `Candidate skill '${uName}' (Depth: ${uDepth}) is a broad parent category of specific requirement '${rName}'. ` +
      `Intersection: '${lcaName}' (Depth: ${lcaDepth}, Graph Distance: ${distance}) → ${scoreFormatted}`
    );
  }

  if (features.same_parent) {
    return (
      `Requirement '${rName}' (Depth: ${rDepth}): Sibling technology match — ` +
      `Candidate skill '${uName}' (Depth: ${uDepth}) shares direct parent '${lcaName}' (Depth: ${lcaDepth}, Graph Distance: ${distance}) with '${rName}' → ${scoreFormatted}`
    );
  }

  if (features.lca_depth >= 2) {
    return (
      `Requirement '${rName}' (Depth: ${rDepth}): Related subdomain match — ` +
      `Candidate skill '${uName}' (Depth: ${uDepth}) and requirement connect through subdomain '${lcaName}' (Depth: ${lcaDepth}, Graph Distance: ${distance}) → ${scoreFormatted}`
    );
  }

  if (features.lca_depth === 1) {
    return (
      `Requirement '${rName}' (Depth: ${rDepth}): Domain-level overlap — ` +
      `Candidate skill '${uName}' (Depth: ${uDepth}) and requirement meet at top-level domain '${lcaName}' (Depth: ${lcaDepth}, Graph Distance: ${distance}) → ${scoreFormatted}`
    );
  }

  return `Requirement '${rName}' (Depth: ${rDepth}): Weak taxonomy match with '${uName}' (Depth: ${uDepth}, Graph Distance: ${distance}) → ${scoreFormatted}`;
}
